agents_infra/agents_infra/utils/communication.py

```diff
@@ -0,0 +1,183 @@
+import logging
+
+from logtools import maybe_log_message
+
+from ..utils.network import (check_http_health, find_first_healthy_url,
+                             is_tcp_reachable, ping_url)
+from ..utils.timestamp import get_current_time
+
+revert_interval = 900
+
+
+class AgentCommunication(object):
+    """
+    Handles communication between an agent and its controller(s),
+    including health checks, failover logic, and posting data.
+    """
+
+    def __init__(self, auth_token_type, post_data_fn, controller_urls):
+        """
+        Initialize AgentCommunication.
+
+        Args:
+            auth_token_type (str): Type of auth token to use in requests.
+            post_data_fn (callable): Function to send data to a controller.
+            controller_urls (list): List of controller URLs in priority order.
+        """
+        self.post_data_fn = post_data_fn
+        self.auth_token_type = auth_token_type
+        self.controller_urls = controller_urls
+        self.current_controller = controller_urls[0]
+
+    @property
+    def logger(self):
+        """
+        Returns a logger named after the agent config or falls back
+        to 'agent-communication'.
+        """
+        name = 'agent-communication'
+        if hasattr(self.post_data_fn.__self__.config, 'name'):
+            name = self.post_data_fn.__self__.config.name
+        return logging.getLogger(name)
+
+    def _switch_controller(self, new_url):
+        """
+        Switch the current controller to a new one.
+        """
+        self.current_controller = new_url
+        self.last_success_time = get_current_time()
+        maybe_log_message(
+            'Controller switched: %s -> %s' % (
+                self.current_controller,
+                new_url
+            ),
+            logger=self.logger,
+            level=logging.INFO
+        )
+
+    def ensure_active_controller(self, api_key):
```

Comment:
Consider thread-safety if this method can be accessed concurrently, since it mutates `self.current_controller`.